jlm/mlir/frontend/MlirToJlmConverter.cpp

```diff
@@ -1189,6 +1189,26 @@ MlirToJlmConverter::ConvertType(const ::mlir::Type & type)
     // RVSDG does not support indices, which are modeled as integers
     return rvsdg::BitType::Create(MlirToJlmConverter::GetIndexBitWidth());
   }
+  else if (auto structType = ::mlir::dyn_cast<::mlir::LLVM::LLVMStructType>(type))
+  {
+    std::vector<std::shared_ptr<const rvsdg::Type>> types;
+    for (auto element : structType.getBody())
+    {
+      types.push_back(ConvertType(element));
+    }
+    auto declaration = jlm::llvm::StructType::Declaration::Create(types);
```

Comment:
You create a new declaration for every new struct type that you get from MLIR, even if it actually represents the same struct type. The equality of the struct type in JLM is determined with the pointer equality of the declaration. Thus, if you want to make sure two struct types are identical, you need to give them the same declaration. Ultimately, this means you need to keep track of unique declarations going from JLM -> MLIR -> JLM